src/openms/source/QC/DBSuitability.cpp

```diff
@@ -0,0 +1,298 @@
+  }
+
+  const std::vector<DBSuitability::SuitabilityData>& DBSuitability::getResults() const
+  {
+    return results_;
+  }
+
+  double DBSuitability::getDecoyDiff_(const PeptideIdentification& pep_id)
+  {
+    double diff = DBL_MAX;
+
+    // get the score of the first two decoy hits
+    double decoy_1 = DBL_MAX;
+    double decoy_2 = DBL_MAX;
+    UInt curr_hit = 0;
+
+    for (const auto& hit : pep_id.getHits())
+    {
+      if (curr_hit > 10) break;
+      ++curr_hit;
+
+      if (!hit.metaValueExists("target_decoy"))
+      {
+        throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No target/decoy information found! Make sure 'PeptideIndexer' is run beforehand."));
+      }
+
+      if (!hit.metaValueExists("MS:1002252"))
+      {
+        throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "No cross correlation score found at peptide hit. Only Comet search engine is supported right now."));
+      }
+
+      if (decoy_1 == DBL_MAX && hit.getMetaValue("target_decoy") == "decoy")
+      {
+        decoy_1 = hit.getMetaValue("MS:1002252");
+        continue;
+      }
+      if (decoy_1 < DBL_MAX && hit.getMetaValue("target_decoy") == "decoy")
+      {
+        decoy_2 = hit.getMetaValue("MS:1002252");
+        break;
+      }
+    }
+
+    if (decoy_2 < DBL_MAX) // if there are two decoy hits
+    {
+      diff = abs(decoy_1 - decoy_2) / pep_id.getHits()[0].getSequence().getMonoWeight(); // normalized by mw
+    }
+
+    // if there aren't two decoy hits DBL_MAX is returned
+    return diff;
+  }
+
+  double DBSuitability::getDecoyCutOff_(const vector<PeptideIdentification>& pep_ids, double reranking_cutoff_percentile)
+  {
+    if (reranking_cutoff_percentile < 0 || reranking_cutoff_percentile > 1)
+    {
+      throw Exception::IllegalArgument(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "'reranking_cutoff_percentile' is not within its allowed range [0,1]. Please select a valid value.");
+    }
+
+    // get all decoy diffs of peptide ids with at least two decoy hits
+    vector<double> diffs;
+    for (const auto& pep_id : pep_ids)
+    {
+      double diff = getDecoyDiff_(pep_id);
+      if (diff < DBL_MAX)
+      {
+        diffs.push_back(diff);
+      }
+    }
+
+    if (double(diffs.size()) / pep_ids.size() < 0.2)
+    {
+      throw(Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "Under 20 % of peptide identifications have two decoy hits. This is not enough for re-ranking. Use the 'no_rerank' flag to still compute a suitability score."));
+    }
+
+    // sort the diffs decreasing and get the percentile one
+    UInt index = round(reranking_cutoff_percentile * diffs.size());
+    
+    if (index >= diffs.size())
+    {
+      return *min_element(diffs.begin(), diffs.end());
+    }
+
+    nth_element(diffs.begin(), diffs.begin() + index, diffs.end(), greater<double>());
```

Comment:
so this is still reversed?
Swap in the next PR?

Reply:
Yes, it is sorted descending. But since I take the actual percentile now and not the (1 - percentile) one like before, this is essentially swapped.

Reply:
My bad, the `greater<double>()` isn't needed anymore.